File: ghostcam-py/ghostcam/provisioning.py
"""First-boot provisioning.

Mirrors camera/provisioning.go. Resolution order for the provision token
and server URL is CLI/env → flat files in data_dir → QR code scan.

QR scanning is platform-gated (real-Pi only via rpicam-still + pyzbar);
on synthetic platforms `scan_qr` returns None, so unprovisioned cameras
in dev/Docker just need the env vars.
"""

from __future__ import annotations

import contextlib
import logging
from pathlib import Path

from ghostcam.client import provision
from ghostcam.config import CameraConfig
from ghostcam.credentials import Credentials, save_credentials
from ghostcam.identity import Identity
from ghostcam.platform import ensure_wifi, scan_qr, wait_for_route

logger = logging.getLogger(__name__)


def _read_trimmed(path: Path) -> str:
    try:
        return path.read_text().strip()
    except OSError:
        return ""


def _resolve_inputs(cfg: CameraConfig) -> tuple[str, str]:
    token = cfg.provision_token
    server_url = cfg.server_url
    if not token:
        token = _read_trimmed(cfg.data_dir / "provision_token")
    if not server_url:
        server_url = _read_trimmed(cfg.data_dir / "server_url")
    if token and server_url:
        return token, server_url
    return "", ""


async def run_provisioning(
    cfg: CameraConfig,
    device_serial: str,
    identity: Identity,
) -> Credentials | None:
    token, server_url = _resolve_inputs(cfg)

    if not token or not server_url:
        qr = await scan_qr()
        if qr is None:
            logger.info("no provision token available, waiting for provisioning")
            return None
        token = qr.t
        server_url = qr.s
        if qr.w:
            try:
                await ensure_wifi(qr.w, qr.p)
            except Exception as e:  # noqa: BLE001
                logger.warning("WiFi from QR failed (ssid=%s): %s", qr.w, e)
            await wait_for_route()

    if not server_url.startswith(("http://", "https://")):
        server_url = "https://" + server_url

    logger.info("attempting provisioning: server=%s", server_url)
    await provision(server_url, token, device_serial, identity)

    creds = Credentials(
        device_id=identity.device_id,
        server_url=server_url,
        identity=identity,
    )
    save_credentials(cfg.data_dir, server_url)

    token_file = cfg.data_dir / "provision_token"
    if token_file.exists():
        with contextlib.suppress(OSError):
            token_file.unlink()

    logger.info("provisioning complete: device_id=%s", creds.device_id)
    return creds

File: legacy_camera/provisioning.go
package main

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
)

// RunProvisioning provisions the camera via a one-time token. Token and server
// URL are resolved in order: CLI/env (cfg.ProvisionToken, cfg.ServerURL) →
// flat files ({dataDir}/provision_token, {dataDir}/server_url) →
// QR code scan via rpicam-still (Linux only, 5-min timeout).
//
// The camera sends its ed25519 public key to the server — no secret is
// returned. Returns nil credentials if no token is available from any source.
func RunProvisioning(ctx context.Context, cfg *CameraConfig, deviceSerial string, identity *Identity) (*Credentials, error) {
	token, serverURL := resolveProvisionInputs(cfg)

	// Fallback: scan QR code from camera
	if token == "" || serverURL == "" {
		qr, err := ScanQR(ctx)
		if err != nil {
			return nil, fmt.Errorf("QR scan: %w", err)
		}
		if qr == nil {
			slog.Info("no provision token available, waiting for provisioning")
			return nil, nil
		}
		token = qr.Token
		serverURL = qr.Server

		// QR may include WiFi credentials
		if qr.WifiSSID != "" {
			psk := qr.WifiPassword
			if err := EnsureWifi(ctx, qr.WifiSSID, &psk); err != nil {
				slog.Warn("WiFi from QR failed", "ssid", qr.WifiSSID, "err", err)
			}
			WaitForRoute(ctx)
		}
	}

	// Ensure server_url is a full URL
	if !strings.HasPrefix(serverURL, "http://") && !strings.HasPrefix(serverURL, "https://") {
		serverURL = "https://" + serverURL
	}

	slog.Info("attempting provisioning", "server", serverURL)

	if err := Provision(ctx, serverURL, token, deviceSerial, identity); err != nil {
		return nil, err
	}

	creds := &Credentials{
		DeviceID:  identity.DeviceID,
		ServerURL: serverURL,
		Identity:  identity,
	}

	if err := SaveCredentials(cfg.DataDir, creds); err != nil {
		return nil, err
	}

	// Remove the one-time token file (if it exists)
	_ = os.Remove(filepath.Join(cfg.DataDir, "provision_token"))

	slog.Info("provisioning complete", "device_id", creds.DeviceID)
	return creds, nil
}

// resolveProvisionInputs returns (token, serverURL) by merging CLI/env
// and flat file sources. Either source can provide either value.
func resolveProvisionInputs(cfg *CameraConfig) (string, string) {
	token := cfg.ProvisionToken
	serverURL := cfg.ServerURL

	if token == "" {
		token = readTrimmedFile(filepath.Join(cfg.DataDir, "provision_token"))
	}
	if serverURL == "" {
		serverURL = readTrimmedFile(filepath.Join(cfg.DataDir, "server_url"))
	}

	if token != "" && serverURL != "" {
		return token, serverURL
	}
	return "", ""
}
